content/learning-paths/mobile-graphics-and-gaming/android_halide/processing-workflow.md

```diff
@@ -0,0 +1,585 @@
+
+Below we show both flavors.
+
+### Tiling with explicit intermediate storage (best for cache efficiency)
+Here we cache gray once per tile so the 3×3 blur can reuse it instead of recomputing RGB -> gray up to 9× per output pixel.
+
+Before using this, remove any earlier compute_root().parallel(y) schedule for blur.
+
+```cpp
+// After defining: input, gray, blur, thresholded
+Halide::Var xo("xo"), yo("yo"), xi("xi"), yi("yi");
+
+// Tile & parallelize the consumer; vectorize inner x on planar output.
+thresholded
+    .tile(x, y, xo, yo, xi, yi, 128, 64)
+    .vectorize(xi, 16)
+    .parallel(yo);
+
+// Compute blur inside each tile and vectorize its inner x.
+blur
+    .compute_at(thresholded, xo)
+    .vectorize(x, 16);
+
+// Cache RGB→gray per tile (reads interleaved input → keep unvectorized).
+gray
+    .compute_at(thresholded, xo)
+    .store_at(thresholded, xo);
+```
+
+In this scheduling:
+* tile(...) splits the image into cache-friendly blocks and makes it easy to parallelize across tiles.
+* blur.compute_at(thresholded, xo) localizes the blur computation to each tile (it doesn’t force storing blur; it just computes it where it’s needed, keeping the working set small).
+* gray.compute_at(...).store_at(...) materializes a tile-local planar buffer for the grayscale intermediate so blur can reuse it within the tile.
+* Vectorization is applied only to planar stages (blur, thresholded), gray stays unvectorized because it reads interleaved input (x-stride = channels).
+
+Recompile your application as before, then run. On our machine, this version ran at ~7.6 FPS (~15.76 MPix/s, ~139 ms/frame), slower than baseline (~12.48 FPS) and the parallelization-only schedule (~14.79 FPS). The 3×3 blur is very small (low arithmetic intensity), the extra writes/reads of a tile-local buffer add overhead, and the interleaved source still limits how efficiently the gray producer can be read/vectorized.
+
+This pattern shines when the cached intermediate is expensive and reused a lot (bigger kernels, multi-use intermediates, or separable/multi-stage pipelines). For a tiny 3×3 on CPU, the benefit often doesn’t amortize.
+
+### Tiling for parallelization (without explicit intermediate storage)
+Tiling can also be used just to partition work across cores, without caching intermediates. This keeps the schedule simple: we split the output into tiles, parallelize across tiles, and vectorize along unit-stride x. Producers are computed inside each tile to keep the working set small, but we don’t materialize extra tile-local buffers:
+```cpp
+// Tiling (partitioning only)
+Halide::Var xo("xo"), yo("yo"), xi("xi"), yi("yi");
+
+thresholded
+    .tile(x, y, xo, yo, xi, yi, 128, 64)  // try 128x64; tune per CPU
+    .vectorize(xi, 16)                    // safe: planar, unit-stride along x
+    .parallel(yo);                        // run tiles across cores
+
+blur
```

Comment:
As shown in fusion.md, blur output and threshold can be simply fused. Instead, what is better to write out is gray stage, either by compute_root or by compute_at.